pkg/server/plugin/noderesolver/aws/iid.go

```diff
@@ -0,0 +1,282 @@
+		return nil, err
+	}
+
+	resp, err := client.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(instanceID)},
+		Filters:     instanceFilters,
+	})
+	if err != nil {
+		return nil, iidError.Wrap(err)
+	}
+
+	selectorSet := map[string]bool{}
+	addSelectors := func(values []string) {
+		for _, value := range values {
+			selectorSet[value] = true
+		}
+	}
+
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			addSelectors(resolveTags(instance.Tags))
+			addSelectors(resolveSecurityGroups(instance.SecurityGroups))
+			if instance.IamInstanceProfile != nil && instance.IamInstanceProfile.Arn != nil {
+				output, err := client.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+					InstanceProfileName: instance.IamInstanceProfile.Arn,
+				})
+				if err != nil {
+					return nil, iidError.Wrap(err)
+				}
+				addSelectors(resolveInstanceProfile(output.InstanceProfile))
+			}
+		}
+	}
+
+	// sort and dedup selectors
+	values := make([]string, 0, len(selectorSet))
+	for value := range selectorSet {
+		values = append(values, value)
+	}
+	sort.Strings(values)
+
+	selectors := new(common.Selectors)
+	for _, value := range values {
+		selectors.Entries = append(selectors.Entries, &common.Selector{
+			Type:  pluginName,
+			Value: value,
+		})
+	}
+
+	return selectors, nil
+}
+
+func (p *IIDResolverPlugin) getRegionClient(region string) (awsClient, error) {
+	// do an initial check to see if p client for this region already exists
+	p.mu.RLock()
+	client, ok := p.clients[region]
+	p.mu.RUnlock()
+	if ok {
+		return client, nil
+	}
+
+	// no client for this region. make one.
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// more than one thread could be racing to create p client (since we had
+	// to drop the read lock to take the write lock), so double check somebody
+	// hasn't beat us to it.
+	client, ok = p.clients[region]
+	if ok {
+		return client, nil
+	}
+
+	if p.conf == nil {
+		return nil, iidError.New("not configured")
+	}
+	p.conf.Region = aws.String(region)
+
+	client, err := p.hooks.newClient(p.conf)
+	if err != nil {
+		return nil, err
+	}
+
+	p.clients[region] = client
+	return client, nil
+}
+
+func resolveTags(tags []*ec2.Tag) []string {
+	values := make([]string, 0, len(tags))
+	for _, tag := range tags {
+		values = append(values, fmt.Sprintf("tag:%s:%s", aws.StringValue(tag.Key), aws.StringValue(tag.Value)))
+	}
+	return values
+}
+
+func resolveSecurityGroups(sgs []*ec2.GroupIdentifier) []string {
+	values := make([]string, 0, len(sgs)*2)
+	for _, sg := range sgs {
+		values = append(values,
+			fmt.Sprintf("sg:id:%s", aws.StringValue(sg.GroupId)),
+			fmt.Sprintf("sg:name:%s", aws.StringValue(sg.GroupName)),
+		)
+	}
+	return values
+}
+
+func resolveInstanceProfile(instanceProfile *iam.InstanceProfile) []string {
+	values := make([]string, 0, len(instanceProfile.Roles))
+	for _, role := range instanceProfile.Roles {
```

Comment:
I think we are missing a check to ensure that instanceProfile is not nil

Reply:
:+1: on defensiveness, will fix.